commons-mongo/src/main/scala/com/avsystem/commons/mongo/typed/TypedMongoCollection.scala

```diff
@@ -0,0 +1,292 @@
+      setupOptions(publisher.asInstanceOf[DistinctPublisher[Any]]).asInstanceOf[DistinctPublisher[BsonValue]]
+
+    Observable.fromReactivePublisher(publisherWithOptions).map(property.format.readBson)
+  }
+
+  def insertOne(
+    value: E,
+    setupOptions: InsertOneOptions => InsertOneOptions = identity
+  ): Task[Unit] =
+    single(nativeCollection.insertOne(value, setupOptions(new InsertOneOptions))).void
+
+  def insertMany(
+    values: Seq[E],
+    setupOptions: InsertManyOptions => InsertManyOptions = identity
+  ): Task[Unit] =
+    single(nativeCollection.insertMany(values.asJava, setupOptions(new InsertManyOptions))).void
+
+  def deleteOne(
+    filter: MongoDocumentFilter[E],
+    setupOptions: DeleteOptions => DeleteOptions = identity
+  ): Task[DeleteResult] =
+    single(nativeCollection.deleteOne(filter.toBson, setupOptions(new DeleteOptions)))
+
+  def deleteMany(
+    filter: MongoDocumentFilter[E],
+    setupOptions: DeleteOptions => DeleteOptions = identity
+  ): Task[DeleteResult] =
+    single(nativeCollection.deleteMany(filter.toBson, setupOptions(new DeleteOptions)))
+
+  def updateOne(
+    filter: MongoDocumentFilter[E],
+    update: MongoDocumentUpdate[E],
+    upsert: Boolean = false, // extracted as separate param because it's very commonly used
+    setupOptions: UpdateOptions => UpdateOptions = identity
+  ): Task[UpdateResult] = {
+    val options = setupOptions(new UpdateOptions).upsert(upsert)
+    val (updateBson, arrayFilters) = update.toBsonAndArrayFilters
+    if (!arrayFilters.isEmpty) {
+      options.arrayFilters(arrayFilters)
+    }
+    single(nativeCollection.updateOne(filter.toBson, updateBson, options))
+  }
+
+  def updateMany(
+    filter: MongoDocumentFilter[E],
+    update: MongoDocumentUpdate[E],
+    upsert: Boolean = false, // extracted as separate param because it's very commonly used
+    setupOptions: UpdateOptions => UpdateOptions = identity
+  ): Task[UpdateResult] = {
+    val options = setupOptions(new UpdateOptions).upsert(upsert)
+    val (updateBson, arrayFilters) = update.toBsonAndArrayFilters
+    if (!arrayFilters.isEmpty) {
+      options.arrayFilters(arrayFilters)
+    }
+    single(nativeCollection.updateMany(filter.toBson, updateBson, options))
+  }
+
+  def replaceOne(
+    filter: MongoDocumentFilter[E],
+    replacement: E,
+    upsert: Boolean = false, // extracted as separate param because it's very commonly used
+    setupOptions: ReplaceOptions => ReplaceOptions = identity
+  ): Task[UpdateResult] = {
+    val options = setupOptions(new ReplaceOptions).upsert(upsert)
+    single(nativeCollection.replaceOne(filter.toBson, replacement, options))
+  }
+
+  def bulkWrite(
+    writes: Seq[MongoWrite[E]],
+    setupOptions: BulkWriteOptions => BulkWriteOptions = identity
+  ): Task[BulkWriteResult] = {
+    val requests = writes.iterator.map(_.toWriteModel).to[JList]
+    single(nativeCollection.bulkWrite(requests, setupOptions(new BulkWriteOptions)))
```

Comment:
You could consider skipping the call if `writes` is empty - the native driver notoriously throws exceptions in such case which is usually not what the user expects.

Reply:
How do I create `BulkWriteResult` to represent that?

Reply:
You probably don't. Since you're wrapping the API, you can always introduce your own result type.

Reply:
But I want to expose `BulkWriteResult` since it contains useful information. Do you want me to wrap it into an `Opt` or something?

Reply:
I wouldn't mind that, basically yes.

Reply:
I don't like this because it would force pattern matching or some other empty `Opt` handling in situations where the user is sure that `writes` are non-empty. Also, this goes against the "wrap with better API but don't change semantics" approach of `TypedMongoCollection`.